test/signal_test.dart

```diff
@@ -0,0 +1,385 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:cli_script/cli_script.dart';
+import 'package:test/test.dart';
+
+import 'util.dart';
+
+void main() {
+  group('on Script', () {
+    test('interrupts with default signal', () async {
+      var script =
+          mainScript('await Future<void>.delayed(Duration(seconds: 100));');
+
+      expect(await script.kill(), true);
+
+      expect(script.done, throwsScriptException(-15));
+    });
+
+    test('interrupts with custom signal', () async {
+      var script =
+          mainScript('await Future<void>.delayed(Duration(seconds: 100));');
+
+      expect(await script.kill(ProcessSignal.sigint), true);
+      expect(script.done, throwsScriptException(-2));
+    });
+
+    test('can interrupt after Future.timeout', () async {
+      var script =
+          mainScript('await Future<void>.delayed(Duration(seconds: 100));');
+
+      script.done.timeout(Duration(milliseconds: 1), onTimeout: script.kill);
+      expect(script.done, throwsScriptException(-15));
+    });
+
+    test("can't be interrupt after it already exited", () async {
+      var script = mainScript('print("done!");');
+
+      expect(await script.output, 'done!');
+      expect(await script.kill(), false);
+    });
+
+    test("can be trapped without interrupting", () async {
+      var script = _watchSignalsAndExit();
+      var output = script.output;
+
+      await Future<void>.delayed(Duration(seconds: 1));
+      expect(await script.kill(), true);
+
+      expect(await output, 'SIGTERM\nbye!');
+      expect(await script.kill(), false);
+    });
+  });
+
+  group('on Script.capture', () {
+    test("does not interrupt by default", () async {
+      var script = Script.capture((_) async {
+        await Future<void>.delayed(Duration(seconds: 3));
```

Comment:
Relying on wall-clock sleeps is a recipe for slow tests and flaky failures. It's often better to use an external signal (like a `Completer`) to explicitly control which futures fire when, but in a pure-Dart test like this you can also use the [`fake_async` package](https://pub.dev/packages/fake_async) to use relative timings without relying on the vicissitudes of real time.